docs/proposals/20200506-conditions.md

```diff
@@ -0,0 +1,550 @@
+type ConditionSeverity string
+
+const (
+  // ConditionSeverityError specifies that a failure of a condition type
+  // should be viewed as an error.
+  ConditionSeverityError ConditionSeverity = "Error"
+
+  // ConditionSeverityWarning specifies that a failure of a condition type
+  // should be viewed as a warning, but that things could still work.
+  ConditionSeverityWarning ConditionSeverity = "Warning"
+
+  // ConditionSeverityInfo specifies that a failure of a condition type
+  // should be viewed as purely informational, and that things could still work.
+  ConditionSeverityInfo ConditionSeverity = "Info"
+
+  // ConditionSeverityNone should apply only if the condition is in state "True".
+  // As ConditionSeverityNone is the default for conditions we use the empty string (coupled with omitempty)
+  ConditionSeverityNone ConditionSeverity = ""
+)
+
+// Condition defines an extension to status (i.e. an observation) of a Cluster API resource.
+type Condition struct {
+   // Type of condition.
+   // +required
+   Type ConditionType `json:"type" description:"type of status condition"`
+
+   // Status of the condition, one of True, False, Unknown.
+   // +required
+   Status corev1.ConditionStatus `json:"status"`
+
+   // Severity with which to treat failures of this type of condition.
+   // When this is not specified, it defaults to Error.
+   // +optional
+   Severity ConditionSeverity `json:"severity,omitempty"`
+
+   // LastTransitionTime is the last time the condition transitioned from one status to another.
+   // +required
+   LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
+
+   // The reason for the condition's last transition.
+   // Reasons should be one-word CamelCase.
+   // +optional
+   Reason string `json:"reason,omitempty" description:"one-word CamelCase reason for the condition's last transition"`
+
+   // A human readable message indicating details about the transition.
+   // +optional
+   Message string `json:"message,omitempty" description:"human-readable message indicating details about last transition"`
+}
+
+// Conditions define an extension to status (i.e. an observation) of a Cluster API resource.
+type Conditions []Condition
+```
+
+Every Cluster API resource except data-only objects which aren't reconciled (i.e. `KubeadmConfigTemplate`)
+SHOULD have conditions field in the status struct. e.g.
+
+```golang
+// ClusterStatus represent the status for the Cluster object
+type ClusterStatus struct {
+    ...
+    // Conditions define a list of readiness conditions for the Cluster object
+    Conditions Conditions `json:"conditions,omitempty"
+} 
+```
+
+Each Cluster API resource SHOULD define its own set of condition types, e.g. 
+
+```golang
+// ConditionTypes for the cluster object 
+const (
+	ClusterInfrastructureReady ConditionType = "ClusterInfrastructureReady"
+	
+	...
+)
+```
+
+Condition types MUST have a consistent polarity (i.e. "True = good");
```

Comment:
While I agree with this in concept, I'm not sure we would want to force this in practice.

In particular, I'm wondering about how this representation would affect the way that we deal with the absence of a given condition, particularly in the case of MachineHealthCheck compared to say something like ClusterInfrastructureReady.

In the case of a missing condition related to the Cluster Infrastructure being ready we would want to consider the default in the case of the condition being not present to be "not ready"(false/unknown) and block any operations that would want Cluster Infrastructure Ready.

However, in the case of MachineHealthCheck if we had a condition that represented an MHC passing (good), we would need to treat the absence of the condition as being true to avoid attempting to perform remediation prior to the MHC running. Not only does this seem a bit counter-intuitive on the remediation side (assuming success in the event of a missing condition) it also seems counter-intuitive from an implementation side for MHC as well, since we cannot really be confident that it is healthy, just that the conditions we are checking for have not failed.

Reply:
Yes, we already are facing this problem in #3056 with the verbal contortions needed to make a condition that means "remediation is needed". We couldn't come up with anything better than `OwnerRemediationSucceeded` being set to false upon failure, which is creating some confusing semantics.

Reply:
@detiber @benmoss 
It is a trade-off.
I'm +1 for trying to enforce this rule because having a consistent polarity ensures a simple and clear contract with the consumers and it allows to manipulate conditions in a simple and consistent way without being forced to implements specific logic for each condition type. 

> I'm wondering about how this representation would affect the way that we deal with the absence of a given condition

IMO a missing condition is unknown (not True, not False); also in this case, a consistent polarity can help in reading code.